Levels unity

danalex97 opened the pull request:
This is a pull request meant to complete and extend the #220 pull request. The new features are support for levels exported from Unity(see https://github.com/ocadotechnology/aimmo-unity/tree/level_generation for more deails). Now both .txt levels and .json levels are supported. 

Another important change is a refactoring for the cells: we can now pass a sprite description metadata formatted as a json. Some parts of the world_map may seem boilerplate, but this is due to the fact that #223 is yet to be merged.

danalex97 commented:
Useful link: https://github.com/ocadotechnology/aimmo/wiki/Levels

CelineBoudier commented:
I would squash (or rename?) a few commits there, to get a clean history if possible. Commits like "Fixed small merge problem." are not really helpful for instance...

CelineBoudier commented on players/views.py:
```diff
@@ -143,7 +143,7 @@ def watch_level(request, num):
     return _render_game(request, game)
 
-
+# This function adds the Levels to the database
```
might be better as a documentation below the title? But the title is explicit enough I think?

danalex97 replied:
:+1:

CelineBoudier commented on aimmo-game/tests/test_simulation/test_map_generator.py:
```diff
@@ -8,7 +8,7 @@
 from simulation.map_generator import get_random_edge_index
 from simulation.world_map import WorldMap
```
I don't think you need this import anymore :)

danalex97 replied:
Indeed it is not.

danalex97 commented:
I squashed some commits and removed some prints. Hopefully now is more clean.

CelineBoudier commented on aimmo-game/simulation/game_state.py:
```diff
@@ -1,6 +1,5 @@
 from simulation.avatar import fog_of_war
 
-
```
two lines please https://www.python.org/dev/peps/pep-0008/#blank-lines

CelineBoudier commented on aimmo-game/simulation/levels/decoders.py:
```diff
@@ -0,0 +1,75 @@
+from simulation.pickups import HealthPickup
+from simulation.pickups import InvulnerabilityPickup
+from simulation.pickups import DamagePickup
+
```
two lines please https://www.python.org/dev/peps/pep-0008/#blank-lines

CelineBoudier commented on aimmo-game/simulation/levels/decoders.py:
```diff
@@ -0,0 +1,75 @@
+    if "sprite" in json:
+        return json["sprite"]
+    return {}
+
```
two lines please https://www.python.org/dev/peps/pep-0008/#blank-lines

CelineBoudier commented on aimmo-game/simulation/levels/decoders.py:
```diff
@@ -0,0 +1,75 @@
+    @abc.abstractmethod
+    def decode(self, json, world_map):
+        pass
+
```
two lines please https://www.python.org/dev/peps/pep-0008/#blank-lines

CelineBoudier commented on aimmo-game/simulation/levels/decoders.py:
```diff
@@ -0,0 +1,75 @@
+
+def get_sprite(json):
+    if "sprite" in json:
+        return json["sprite"]
```
please use https://docs.python.org/2/library/stdtypes.html#dict.get to get the value here :)

CelineBoudier commented on aimmo-game/simulation/levels/decoders.py:
```diff
@@ -0,0 +1,75 @@
+        x, y = int(json["x"]), int(json["y"])
+        world_map = WorldMapStaticSpawnDecorator(world_map, Location(x, y))
+        world_map.get_cell(Location(x, y)).cell_content = ScoreLocation(get_sprite(json))
+
```
two lines please https://www.python.org/dev/peps/pep-0008/#blank-lines

CelineBoudier commented on aimmo-game/simulation/levels/decoders.py:
```diff
@@ -0,0 +1,75 @@
+    def decode(self, json, world_map):
+        x, y = int(json["x"]), int(json["y"])
+        world_map.get_cell(Location(x, y)).cell_content = Obstacle(get_sprite(json))
+
```
two lines please https://www.python.org/dev/peps/pep-0008/#blank-lines

CelineBoudier commented on aimmo-game/simulation/levels/decoders.py:
```diff
@@ -0,0 +1,75 @@
+            world_map.get_cell(Location(x, y)).pickup = HealthPickup(world_map.get_cell(Location(x, y)), int(json["health_restored"]))
+        if json["type"] == "damage":
+            world_map.get_cell(Location(x, y)).pickup = DamagePickup(world_map.get_cell(Location(x, y)))
+
```
two lines please https://www.python.org/dev/peps/pep-0008/#blank-lines

CelineBoudier commented on aimmo-game/simulation/levels/levels.py:
```diff
@@ -0,0 +1,85 @@
+_SCRIPT_LOCATION = os.path.abspath(os.path.dirname(__file__))
+_MAPS_FOLDER = os.path.join(_SCRIPT_LOCATION, "maps")
+_JSON_FOLDER = os.path.join(_SCRIPT_LOCATION, "json")
+
```
two lines please https://www.python.org/dev/peps/pep-0008/#blank-lines

CelineBoudier commented on aimmo-game/simulation/levels/levels.py:
```diff
@@ -0,0 +1,85 @@
+
+    def generate_json(self):
+        return self.parser.get_objects()
+
```
two lines please https://www.python.org/dev/peps/pep-0008/#blank-lines

CelineBoudier commented on aimmo-game/simulation/levels/parsers.py:
```diff
@@ -0,0 +1,165 @@
+import re
+
+from transforms import CellTransform
+
```
two lines please https://www.python.org/dev/peps/pep-0008/#blank-lines

CelineBoudier commented on aimmo-game/simulation/levels/parsers.py:
```diff
@@ -0,0 +1,165 @@
+def call_method(o, name):
+    x = getattr(o, name)()
+    return x
+
```
two lines please https://www.python.org/dev/peps/pep-0008/#blank-lines

CelineBoudier commented on aimmo-game/simulation/levels/parsers.py:
```diff
@@ -0,0 +1,165 @@
+    def get_objects(self):
+        self.parse_json()
+        return self.map
+
```
two lines please https://www.python.org/dev/peps/pep-0008/#blank-lines

CelineBoudier commented on aimmo-game/simulation/levels/transforms.py:
```diff
@@ -0,0 +1,29 @@
+import hashlib
+
```
two lines please https://www.python.org/dev/peps/pep-0008/#blank-lines

CelineBoudier commented on aimmo-game/simulation/map_generator.py:
```diff
@@ -13,47 +13,52 @@
-    def get_game_state(self, avatar_manager):
-        return GameState(self.get_map(), avatar_manager, self.check_complete)
+LOGGER = logging.getLogger(__name__)
 
```
you get the idea :)

danalex97 replied:
Just to check, two lines please https://www.python.org/dev/peps/pep-0008/#blank-lines?

CelineBoudier replied:
yes :D

CelineBoudier commented on players/views.py:
```diff
@@ -143,7 +143,6 @@ def watch_level(request, num):
     LOGGER.debug('Displaying game with id %s', game.id)
     return _render_game(request, game)
 
-
```
why?

CelineBoudier commented on all_tests.py:
```diff
@@ -15,7 +15,6 @@
 BASE_DIR = os.path.abspath(os.path.dirname(__file__))
 APPS = ('', 'aimmo-game/', 'aimmo-game-worker/', 'aimmo-game-creator/')
 
-
```
?

danalex97 commented:
I did the changes, for some reason Git does not show the comments outdated for enters.

CelineBoudier commented on aimmo-game/simulation/levels/completion_checks.py:
```diff
@@ -0,0 +1,16 @@
+
+    return main_avatar.score > 1000
+
+COMPLETION_CHECKS = {
```
can't that be templated somehow and linked to the max number of levels defined somewhere?

danalex97 replied:
The idea is that if I remove the completion check, it will just use the default one, that is `return False`.

CelineBoudier commented on aimmo-game/simulation/custom_map.py:
```diff
@@ -0,0 +1,183 @@
+        return GameState(self.get_map(), avatar_manager, self.check_complete)
+
+    def check_complete(self, game_state):
+        return False
```
huh?

danalex97 replied:
The default behavior for the check complete is to to never end the level. We override that in levels with the specific checks.

CelineBoudier replied:
try abstracting it?

OlafSzmidt replied:
@CelineBoudier  @danalex97  I'm not sure about python as it's still quite new to me, but I found the readibility of abstract methods/classes really difficult. Maybe we should use something like '@abc.abstractmethod'?

https://pymotw.com/3/abc/

danalex97 replied:
Yeah, that's how I generally do it. But I wanted to make that a method with a default behavior, as all the classes EmptyMapGenerator, Main and JSONLevelGenerator use check_complete. I think is better this way.

Why we have the parameter game state is that we might want some custom completion checks that do something like check the score and so on.

CelineBoudier commented on aimmo-game/tests/test_simulation/test_custom_map.py:
```diff
@@ -0,0 +1,103 @@
+class TestAutomaticLevelGenerators(unittest.TestCase):
+    def test_generate_level_class(self):
+        def completed_check(self, y):
+            return False
```
same, why a parameter here if it's only to return False

danalex97 replied:
I think the only reason is to fit the interface.

OlafSzmidt commented:
Any updates on merging this?

danalex97 commented:
This was reviewed at some point, but then I rebased. @CelineBoudier reviewed this besides my last commit, As I rebased, git automatically dismissed the stale review. 

Now there seem to be some conflicts. @OlafSzmidt I can rebase again or merge and someone can review this afterwards...

danalex97 commented:
@OlafSzmidt Maybe we should rebase this after you finish the refactoring for MLP?

mrniket commented:
I'm gonna close this as we are redoing it to work with 3D rendering
